src/nupic/encoders/RandomDistributedScalarEncoder.hpp

```diff
@@ -0,0 +1,799 @@
+* published by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+* See the GNU Affero Public License for more details.
+*
+* You should have received a copy of the GNU Affero Public License
+* along with this program.  If not, see http://www.gnu.org/licenses.
+*
+* http://numenta.org/licenses/
+* ---------------------------------------------------------------------
+*/
+
+#ifndef NTA_randomdistributedscalarencoder_HPP
+#define NTA_randomdistributedscalarencoder_HPP
+
+#include <cstring>
+#include <iostream>
+#include <string>
+#include <vector>
+#include <tuple>
+#include <map>
+#include <list>
+#include <array>         
+#include <random>        
+#include <chrono>
+
+#include "Encoder.hpp"
+#include "nupic/utils/Random.hpp" 
+
+using namespace std;
+using namespace nupic;
+
+namespace encoders { 
+
+	/**
+	* <p>
+	* A scalar encoder encodes a numeric (floating point) value into an array of
+	* bits.
+	*
+	* This class maps a scalar value into a random distributed representation that
+	* is suitable as scalar input into the spatial pooler. The encoding scheme is
+	* designed to replace a simple ScalarEncoder. It preserves the important
+	* properties around overlapping representations. Unlike ScalarEncoder the min
+	* and max range can be dynamically increased without any negative effects. The
+	* only required parameter is resolution, which determines the resolution of
+	* input values.
+	*
+	* Scalar values are mapped to a bucket. The class maintains a random
+	* distributed encoding for each bucket. The following properties are maintained
+	* by RandomDistributedEncoder:
+	* </p>
+	* <ol>
+	* <li>Similar scalars should have high overlap. Overlap should decrease
+	* smoothly as scalars become less similar. Specifically, neighboring bucket
+	* indices must overlap by a linearly decreasing number of bits.
+	*
+	* <li>Dissimilar scalars should have very low overlap so that the SP does not
+	* confuse representations. Specifically, buckets that are more than w indices
+	* apart should have at most maxOverlap bits of overlap. We arbitrarily (and
+	* safely) define "very low" to be 2 bits of overlap or lower.
+	*
+	* Properties 1 and 2 lead to the following overlap rules for buckets i and j:<br>
+	*
+	* <pre>
+	* {@code
+	* If abs(i-j) < w then:
+	* 		overlap(i,j) = w - abs(i-j);
+	* else:
+	* 		overlap(i,j) <= maxOverlap;
+	* }
+	* </pre>
+	*
+	* <li>The representation for a scalar must not change during the lifetime of
+	* the object. Specifically, as new buckets are created and the min/max range is
+	* extended, the representation for previously in-range scalars and previously
+	* created buckets must not change.
+	* </ol>
+	*
+	*
+	* @author Numenta
+	* @author Anubhav Chaturvedi
+	*/
+	template<typename T>
+	class RandomDistributedScalarEncoder : public Encoder<T>
```

Comment:
I have an alternative implementation of this class started in PR #278

It uses a similar algorithm.

Reply:
same as https://github.com/htm-community/nupic.cpp/pull/291#discussion_r261834505 
We'll remove one of them.